### PR TITLE
[WIP] Docker image Caching

### PR DESCRIPTION
**Pull request by salamb**
Manage Docker image meta data by storing it in the mongo database. This allows for the rest endpoints to be generated upon startup quickly by not requiring docker container executions to get cli information for each image. This PR also provides a rest endpoint to add or remove docker images

### CODE REVIEW

**Comment by salamb** [comment]
@manthey @cdeepakroy @zachmullen  
PTAL

**Comment by salamb** [comment]
Some notes:
The slicer-xmlspec is not being checked for structural validity due to a problem in the schema for slicer cli modules, I will make an issue on the slicer_module issue once I confirm it is not my usage of lxml

The DockerImage schema allows additional properties to permit dictionaries containing the _id element from the mongo client. 
 
Currently it is assumed that the cli names provided are safe to use as part of the restroute, will add a regex to only accept alpha numeric names only, or I can strip the cli name of bad characters and use the massaged cli name for the rest route

File: server/docker_resource.py [hunk 1]
```diff
@@ -56,382 +58,237 @@ def __init__(self, name):
...
-        for image in current.getDockerImageList():
-            data[image] = current.getCLIListSpec(image)
+        for val in cache:
+            name, tag, imgData = self.createRestDataForImageVersion(val)
+            if name in data:
+                data[name][tag] = imgData
+            else:
+                data[name] = {}
+                data[name][tag] = imgData
+
```
Comment (manthey) [inline server/docker_resource.py]: You could replace lines 74 - 79 with
```
            data.setdefault(name, {})[tag] = imgData
```

File: server/models/docker_image.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        """
+        return [img.name
+                for (imgHash, img) in iteritems(self.data)]
+
+    def getImages(self):
+        """
+        Get a list of Docker images objects stored in the cache
+        """
+        return [img
+                for (imgHash, img) in iteritems(self.data)]
```
Comment (manthey) [inline server/models/docker_image.py]: More simply, `return list(self.data.values())`

File: server/models/docker_image.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        except Exception as err:
+            raise DockerImageError('Failed to add the '
+                                   'docker image to the cache' + err.__str__())
+
+    def getDockerImageList(self):
+        """
+        Get the list docker image names in the cache
+        """
+        return [img.name
+                for (imgHash, img) in iteritems(self.data)]
```
Comment (manthey) [inline server/models/docker_image.py]: More simply: `return [img.name for img in self.data.values()]`

File: server/models/docker_image.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        This check does not use the docker image id, and therefore will not
+        treat two equivalent images with different names as similar
+        :param name: The docker image name
+        :type name:string
+        """
+        imageKey = self._getHashKey(name)
+        if imageKey in self.data:
+            return True
+        else:
+            return False
```
Comment (manthey) [inline server/models/docker_image.py]: More simply, instead of the if / else, `return imageKey in self.data`

File: server/docker_resource.py [hunk 5]
```diff
@@ -56,382 +58,237 @@ def __init__(self, name):
...
             raise RestException('name was not a valid JSON list or string.')
 
-    def _deleteImage(self, names):
+        self._deleteImage(nameList, deleteImage)
+
+    def _deleteImage(self, names, deleteImage):
         """
-        Removes the docker images and there respective clis from the settings
+        Removes the docker images and there respective clis endpoints
         :param name: The name of the docker image (user/rep:tag)
```
Comment (manthey) [inline server/docker_resource.py]: Can you add the description of deleteImage to the docstring?

**Comment by manthey** [comment]
In general, I'd rather log exceptions and errors than just print them.  We can use girder's logging methods: either `from girder import logger`, or the brand new `from girder import logprint` if you want to *also* print to stdout.  Then, instead of `print(foo)`, either use `logger.error(foo)` or `logprint.error(foo)`, depending on what you chose.  Exceptions can get reported automatically: instead of code like:
```
   except Exception as e:
        print('Failed to create REST endpoints for %s: %s' % (cliRelPath, e))
```
you can use
```
    except Exception:
        logprint.exception('Failed to create REST endpoints for %s' % cliRelPath)
```
and the exception will be formatted and included in the log.

Logging has the virtual that we can poll the logs through the swagger interface rather than having to log in to the actual machine and find Girder's stdout.

File: server/models/docker_image.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+                self.name = self.data[DockerImage.imageName]
+                self.hash = DockerImage.getHashKey(self.name)
+            else:
+
+                raise DockerImageError('Image should be a string, or dict'
+                                       ' could not add the image',
+                                       'bad init val')
+        except Exception as err:
+                raise DockerImageError('Could not initialize instance'
+                                       ' of Docker Image \n'+err.__str__())
```
Comment (manthey) [inline server/models/docker_image.py]: In python the `__str__` method is called when you ask for a object's string representation. `str(err)` is preferred to `err.__str__()`.  Moreover, you can use printf-style formatting: `raise DockerImageError('Could not initialize instance of Docker Image: %s' % err)`, which internally will invoke the `__str__` method.

File: server/rest_slicer_cli.py [hunk 7]
```diff
@@ -973,59 +972,57 @@ def genRESTEndPointsForSlicerCLIsInDockerCache(info, restResource, dockerCache):
...
-    if isinstance(restResource, str):
-        restResource = type(restResource,
-                            (Resource, ),
-                            {'resourceName': restResource})()
-
-    restResourceName = type(restResource).__name__
-
-    # Add REST routes for slicer CLIs in each docker image
-    cliList = []
+        raise Exception('restResource must either be a '
```
Comment (manthey) [inline server/rest_slicer_cli.py]: Missing `string or `

File: server/models/docker_image.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+
+class DockerImageNotFoundError(DockerImageError):
+    def __init__(self, message, image_name, locations=[]):
+        super(DockerImageNotFoundError, self).__init__(message, image_name)
+        # list of registries tried(local dockerhub etc )
+        self.locations = locations
+
+
+class DockerImageDataError(DockerImageError):
```
Comment (manthey) [inline server/models/docker_image.py]: It looks like this error is no longer in use anywhere.

File: server/models/docker_image.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+        else:
+            return False
+
+    def getAllCliSpec(self):
+        spec_dict = {}
+
+        for (key, val) in iteritems(self.data):
+            spec_dict[val.name] = val.getCLIListSpec()
+
+        return spec_dict
```
Comment (manthey) [inline server/models/docker_image.py]: The code from line 200 to here is largely uncovered by the tests.  If it isn't too much work, I'd love to see a test for it.

File: server/models/dockerimagemodel.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+from ..models import DockerImage, DockerImageError, \
+    DockerImageNotFoundError, DockerCache, DockerImageStructure
+
+# from six import iteritems
+# import os
+# from lxml import etree
+# from StringIO import StringIO
+
+
+class Dockerimagemodel(AccessControlledModel):
```
Comment (cdeepakroy) [inline server/models/dockerimagemodel.py]: rename to `DockerImageModel` to follow pep 8 conventions for class names

Comment (manthey) [inline server/models/dockerimagemodel.py]: Girder uses a function called `camelcase` to convert between a model name and the model's class name.  If the model name if `dockerimagemodel`. the class has to be `Dockerimagemodel`.  If, instead, the model were renamed to `docker_image_model`, then the class name is `DockerImageModel`.  To rename the class, you also have to rename the model.

File: server/models/dockerimagemodel.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        nonExist = []
+        img_list = self._getAll()
+        dockerCache = DockerCache()
+        for img in img_list:
+            try:
+                self._ImageExistsLocally(img.name)
+                dockerCache.addImage(img)
+            except DockerImageNotFoundError:
+                nonExist.append(img.name)
+        self.removeImages(nonExist)
```
Comment (cdeepakroy) [inline server/models/dockerimagemodel.py]: After `nonExist = []` in 175, it is not being set anywhere?

File: server/models/docker_image.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+            if isinstance(img, DockerImage):
+                self.data[img.hash] = DockerImage(img.getRawData())
+            else:
+                raise DockerImageError('Tried to add a non '
+                                       'docker image object to cache')
+        except Exception as err:
+            raise DockerImageError('Failed to add the '
+                                   'docker image to the cache' + err.__str__())
+
+    def getDockerImageList(self):
```
Comment (cdeepakroy) [inline server/models/docker_image.py]: Rename to `getDockerImageNames`

File: server/models/docker_image.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        """
+        Get the list docker image names in the cache
+        """
+        return [img.name
+                for (imgHash, img) in iteritems(self.data)]
+
+    def getImages(self):
```
Comment (cdeepakroy) [inline server/models/docker_image.py]: Rename to `getDockerImages` or `getDockerImageList`

File: server/models/docker_image.py (hunk 12 above)
Comment (cdeepakroy) [inline server/models/docker_image.py]: Rename to `getImageNames`?

File: server/models/docker_image.py [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        """
+        Get a list of Docker images objects stored in the cache
+        """
+        return [img
+                for (imgHash, img) in iteritems(self.data)]
+
+    def getImage(self, name):
```
Comment (cdeepakroy) [inline server/models/docker_image.py]: Rename to `getImageByName`

File: plugin_tests/docker_test.py [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+            imgVersions = data[userAndRepo]
+            self.assertHasKeys(imgVersions, [tag])
+
+    def getEndpoint(self):
+        resp = self.request(path='/HistomicsTK/HistomicsTK/docker_image',
+                            user=self.admin)
+        self.assertStatus(resp, 200)
+        return json.loads(self.getBody(resp))
+
+    def noImages(self):
```
Comment (cdeepakroy) [inline plugin_tests/docker_test.py]: rename to `assertNoImages`

File: plugin_tests/docker_test.py [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+        self.noImages()
+        self.addImage(img_name, JobStatus.SUCCESS)
+        self.imageIsLoaded(img_name, True)
+        self.deleteImage(img_name, True, True, JobStatus.SUCCESS)
+        try:
+            from docker import Client
+            docker_client = Client(base_url='unix://var/run/docker.sock')
+
+        except Exception as err:
+            self.fail('could not create the docker client '+err.__str__())
```
Comment (cdeepakroy) [inline plugin_tests/docker_test.py]: You may want to move the loading of `docker_client` into the setup function so multiple test functions could use it if they want.

**Comment by cdeepakroy** [comment]
@salamb Few comments. Rest LGTM!

File: server/rest_slicer_cli.py (hunk 7 above)
Comment (salamb) [inline server/rest_slicer_cli.py]: The resource has to be a Docker Resource only

if not isinstance(restResource, DockerResource):
        raise Exception('restResource must be a '
                        'Docker Resource')

File: server/models/docker_image.py (hunk 9 above)
Comment (salamb) [inline server/models/docker_image.py]: As the use cases changed,(handling multiple versions of the same image etc) and multiple design changes,  these methods are no longer used. I was considering removing them if I could not see a future use case for them.

**Comment by salamb** [comment]
To increase test coverage I can attempt to add an image that is not histomicstk related (docker hello world) just to verify graceful and informative termination of the girder job

**Comment by cdeepakroy** [comment]
@manthey Can we merge this? 

Me and bilal are working on moving this functionality into [girder/slicer_cli](https://github.com/girder/slicer_cli) so other projects have access to it.

**Comment by manthey** [comment]
LGTM.  I'll add an issue to convert print statements to send information to logs.

**Comment by cdeepakroy** [comment]
Ok!

**Pull request merged**